network.go

```diff
@@ -237,6 +279,159 @@ func getLinkByName(netHandle *netlink.Handle, name string, expectedLink netlink.
 	return nil, fmt.Errorf("Incorrect link type %s, expecting %s", link.Type(), expectedLink.Type())
 }
 
+func xconnectVMNetwork(netPair *NetworkInterfacePair, connect bool) error {
+	switch DefaultNetInterworkingModel {
+	case ModelBridged:
+		netPair.NetInterworkingModel = ModelBridged
+		if connect {
+			return bridgeNetworkPair(netPair)
+		} else {
+			return unBridgeNetworkPair(*netPair)
+		}
+	case ModelMacVtap:
+		netPair.NetInterworkingModel = ModelMacVtap
+		if connect {
+			return tapNetworkPair(netPair)
+		} else {
+			return untapNetworkPair(*netPair)
+		}
+	case ModelEnlightened:
+		return fmt.Errorf("Unsupported networking model")
+	default:
+		return fmt.Errorf("Invalid networking model")
+	}
+}
+
+func cleanupFds(fds []*os.File, numFds int) {
+
+	maxFds := len(fds)
+
+	if numFds < maxFds {
+		maxFds = numFds
+	}
+
+	for i := 0; i < maxFds; i++ {
+		_ = fds[i].Close()
+	}
+}
+
+func createMacvtapFds(name string, queues int) ([]*os.File, error) {
+
+	fds := make([]*os.File, queues)
+
+	ifIndexPath := path.Join("/sys/class/net", name, "ifindex")
+	fip, err := os.Open(ifIndexPath)
+	if err != nil {
+		return nil, err
+	}
+	defer func() { _ = fip.Close() }()
+
+	scan := bufio.NewScanner(fip)
+	if !scan.Scan() {
+		return nil, fmt.Errorf("Unable to read tap index")
+	}
+
+	i, err := strconv.Atoi(scan.Text())
+	if err != nil {
+		return nil, err
+	}
+
+	//mq support
+	for q := 0; q < queues; q++ {
+
+		tapDev := fmt.Sprintf("/dev/tap%d", i)
+
+		f, err := os.OpenFile(tapDev, os.O_RDWR, 0666)
```

Comment:
Could you define the perms as a `const` at the top of the file? Also, does it absolutely have to be `0666`?

Reply:
Will fix this. What perms do we use for all the sockets etc we create in CC runtime?

Reply:
The proxy uses `0660`:

- https://github.com/clearcontainers/proxy/blob/master/proxy.go#L655

... but for hypervisor sockets we rely on the defaults provided by qemu fwics (there doesn't appear to be a way to request particular perms from qemu for `chardev` devices). Taking a look at a pod:

```
$ sudo ls -ld /run/virtcontainers                                                                     
d--------- 3 root root 60 Oct  9 09:09 /run/virtcontainers
$ sudo ls -ld /run/virtcontainers/pods
d--------- 4 root root 80 Oct  9 11:15 /run/virtcontainers/pods
$ sudo ls -ld /run/virtcontainers/pods/
total 0
d--------- 3 root root 220 Oct  9 11:15 2128c38aa804ac73045e32c737914a6e6c36e33daa80ba6669d989a72e81652b
d--------- 3 root root 220 Oct  9 09:09 412db0928eccbe5a180a5e4be7577457d25fa845b3504129fec014ab89ae375e
$ sudo ls -l /run/virtcontainers/pods/2128c38aa804ac73045e32c737914a6e6c36e33daa80ba6669d989a72e81652b
total 8
d--------- 2 root root 100 Oct  9 11:15 2128c38aa804ac73045e32c737914a6e6c36e33daa80ba6669d989a72e81652b
srwxr-x--- 1 root root   0 Oct  9 11:15 console.sock
srwxr-x--- 1 root root   0 Oct  9 11:15 ctrl.sock
srwxr-x--- 1 root root   0 Oct  9 11:15 hyper.sock
-rw-r--r-- 1 root root   0 Oct  9 11:15 lock
srwxr-x--- 1 root root   0 Oct  9 11:15 monitor.sock
-rw-r--r-- 1 root root 514 Oct  9 11:15 network.json
-rw-r--r-- 1 root root 149 Oct  9 11:15 state.json
srwxr-x--- 1 root root   0 Oct  9 11:15 tty.sock
$ 
```

Aside: I wonder if we want to drop read perms from those state files (and `lock`).

/cc @grahamwhaley, @sameo.